### PR TITLE
Artifacts AWS S3 integration

### PR DESCRIPTION
**Pull request by WengerK**
suggestion of ameliorations

- add `--help`
- add `--dir` parameter instead of `ARTIFACTS_LOG_DIR`

to use artifact, please add the following keys to your `docker-compose.override.yml`
```
      ARTIFACTS_KEY: (see 1password)
      ARTIFACTS_SECRET: (see 1password)
      ARTIFACTS_BUCKET: (see 1password)
      ARTIFACTS_LOG_DIR: /var/www/log
```

You may not use the `ARTIFACTS_LOG_DIR` and use the option `--local-dir`

### CODE REVIEW

File: scripts/docker-as-artifacts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+# Artifacts variables
+export ARTIFACTS_DEST=${ARTIFACTS_DEST:="/usr/bin/artifacts"}
+export ARTIFACTS_KEY=${ARTIFACTS_KEY:=""}
+export ARTIFACTS_SECRET=${ARTIFACTS_SECRET:=""}
+export ARTIFACTS_BUCKET=${ARTIFACTS_BUCKET:="codeship-artifact"}
+export ARTIFACTS_REGION=${ARTIFACTS_REGION:="eu-west-1"}
+
+# Download Artifacts installer.
+curl -sL -o "${ARTIFACTS_INSTALL}" \
```
Comment (gido) [inline scripts/docker-as-artifacts]: can we pre-install this binary at build time (`Dockerfile`) ?

File: scripts/docker-as-artifacts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+# Collect all files, then to upload them one by one.
+files="$(find -L ${ARTIFACTS_LOG_DIR} -type f)"
+echo "$files" | while read file; do
+  # Upload to Amazon S3.
+  "${ARTIFACTS_DEST}" upload \
+    --target-paths "${CI_REPO_NAME}/${CI_BUILD_NUMBER}/log" \
+    "${file}"
+done
+
```
Comment (gido) [inline scripts/docker-as-artifacts]: remove leading EOL

File: scripts/docker-as-artifacts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+: ${ARTIFACTS_DEFAULT_DIR:="/var/www/log"}
+: ${ARTIFACTS_DEFAULT_S3:="${CI_REPO_NAME}/${CI_BUILD_NUMBER}"}
+
+# Artifacts variables
+: ${ARTIFACTS_BIN:="/usr/bin/artifacts"}
+export ARTIFACTS_KEY=${ARTIFACTS_KEY:=""}
+export ARTIFACTS_SECRET=${ARTIFACTS_SECRET:=""}
+export ARTIFACTS_BUCKET=${ARTIFACTS_BUCKET:="codeship-artifact"}
+export ARTIFACTS_REGION=${ARTIFACTS_REGION:="eu-west-1"}
+
```
Comment (gido) [inline scripts/docker-as-artifacts]:
```suggestion
if [ -z "$ARTIFACTS_KEY" ] || [ ! -z "$ARTIFACTS_SECRET" ]; then
  echo "Error: You need to define env variables ARTIFACTS_KEY and ARTIFACTS_SECRET with S3 credentials"
  exit 1
fi

```

File: scripts/docker-as-artifacts (hunk 1 above)
Comment (yannlugrin) [reply scripts/docker-as-artifacts]: Yes, we can install it at build time, could be better.

File: scripts/docker-as-artifacts [hunk 4]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env bash
+set -e
+
+: ${ARTIFACTS_INSTALL:="/usr/bin/artifacts-install"}
+: ${ARTIFACTS_DEFAULT_DIR:="/var/www/log"}
+: ${ARTIFACTS_DEFAULT_S3:="${CI_REPO_NAME}/${CI_BUILD_NUMBER}"}
+
+# Artifacts variables
+: ${ARTIFACTS_BIN:="/usr/bin/artifacts"}
+export ARTIFACTS_KEY=${ARTIFACTS_KEY:=""}
```
Comment (gido) [inline scripts/docker-as-artifacts]: not sure if we need to export these variables

Comment (yannlugrin) [reply scripts/docker-as-artifacts]: should be tested, but it's possible that is required.

File: scripts/docker-as-artifacts (hunk 1 above)
Comment (WengerK) [reply scripts/docker-as-artifacts]: How to do that ? Should I have a param `skip-upload` and use it during build time ? And where should I put the code to install stuff on Build time ?

File: scripts/docker-as-artifacts (hunk 3 above)
Comment (WengerK) [reply scripts/docker-as-artifacts]: We could use color as we did on `docker-as-drupal` ?

```
printf "\e[01;31mError: You need to define env variables ARTIFACTS_KEY and ARTIFACTS_SECRET with S3 credentials.\e[0m\n"
```

File: scripts/docker-as-artifacts (hunk 4 above)
Comment (WengerK) [reply scripts/docker-as-artifacts]: yes we have to, that what artifacts need to works, every `export` is needed by Artifacts itself.

File: scripts/docker-as-artifacts (hunk 1 above)
Comment (gido) [reply scripts/docker-as-artifacts]: @WengerK here: https://github.com/antistatique/docker-php-dev/blob/dev/Dockerfile#L98

I'm not sure to exactly remember how to test the image without publishing it. @yannlugrin can you document this ?

File: scripts/docker-as-artifacts (hunk 4 above)
Comment (gido) [reply scripts/docker-as-artifacts]: I'm pretty sure they will be accessible for every processes when we put them on `codeship-services.yml` (or more precisely, in the [`encrypted_env_file`](https://documentation.codeship.com/pro/builds-and-configuration/environment-variables/#encrypted-environment-variables) in this case)

File: scripts/docker-as-artifacts (hunk 1 above)
Comment (yannlugrin) [reply scripts/docker-as-artifacts]: We have to move this code in `Dockerfile`:

```bash
# Download Artifacts installer.
curl -sL -o "${ARTIFACTS_INSTALL}" \
  "https://raw.githubusercontent.com/travis-ci/artifacts/master/install"

# Make it executable & install Artifacts.
chmod +x "${ARTIFACTS_INSTALL}"
"${ARTIFACTS_INSTALL}"
```

to something like that:
```Dockerfile
RUN set -ex; \
  \
  curl -fsSL  https://raw.githubusercontent.com/travis-ci/artifacts/master/install | bash
```

Could be added in the big install command.

File: scripts/docker-as-artifacts (hunk 3 above)
Comment (yannlugrin) [reply scripts/docker-as-artifacts]: True about `[ ! -z "$ARTIFACTS_SECRET" ]`

File: scripts/docker-as-artifacts (hunk 4 above)
Comment (WengerK) [reply scripts/docker-as-artifacts]: do you have any clue what we should do then ?

File: scripts/docker-as-artifacts (hunk 1 above)
Comment (WengerK) [reply scripts/docker-as-artifacts]: we also need `ARTIFACTS_DEST` into https://raw.githubusercontent.com/travis-ci/artifacts/master/install

**Pull request merged**